### PR TITLE
Add notes about horizontal scaling

### PR DESCRIPTION
**Pull request by sunaurus**
A collection of notes that will hopefully be useful to any other admins attempting to scale horizontally

### CODE REVIEW

File: src/administration/horizontal_scaling.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+practices. If you don't already know about concepts like load balancing,
+object storage, and reverse proxies, then this document will probably be hard (or impossible) to understand.
+
+**Note: scaling Lemmy this way is not necessary, and for small instances, the added overhead most likely
+outweighs any benefits.**
+
+## Why scale horizontally?
+
+- If one of your Lemmy servers dies for any reason, your instance can still remain operational
+- Having multiple Lemmy servers allows you to do rolling upgrades (no downtime needed to upgrade Lemmy!)
```
Comment (phiresky) [inline src/administration/horizontal_scaling.md:14]: Maybe add a qualifier here about that wmight cause really weird or dangerous issues if the new version has larger database migrations ?

edit: i see you have info about it further below

File: src/administration/horizontal_scaling.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+#### Rolling upgrades
+
+In general, there is nothing preventing you from running multiple load balanced Lemmy-ui servers in parallel, but
+without some custom configuration, **rolling upgrades will break Lemmy for your users!** This is because by default,
+Lemmy-ui will serve static files through the express process.
+
+Consider the scenario where you have 2 Lemmy-ui servers, one is running on version 1, and the other has just come online
+running version 2. A user opening your Lemmy in their browser might have their front page html request routed to version 1.
+This html will contain a reference to many other static files which are specific to version 1. Most likely, some of these
+requests will get routed to the server running version 2. **The files will not exist on this server, the server will
```
Comment (phiresky) [inline src/administration/horizontal_scaling.md:35]: I would always recommend turning on ip hash load leveling. that should prevent most of this?

Comment (Nutomic) [reply src/administration/horizontal_scaling.md:35]: Seems like the problem should be prevented by [this change](https://github.com/LemmyNet/lemmy-ui/pull/1684). Or is it not working?

File: src/administration/horizontal_scaling.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+- Horizontal scaling provides additional flexibility in how you upgrade your infrastructure
+- As of Lemmy version 0.18.2, Lemmy appears to be able to use the same resources more efficiently if they are split between multiple Lemmy processes
+
+## Breakdown of Lemmy components
+
+### Lemmy-ui
+
+Lemmy-ui is the main frontend for Lemmy. It consists of an expressjs based server-side process (necessary for SSR) and
+client code which run in the browser. It does not use a lot of resources and will happily run on quite low powered
+servers.
```
Comment (Nutomic) [inline src/administration/horizontal_scaling.md:24]: This general breakdown of components could fit well into the main administration page. No need to hide it here as its also useful for admins who dont do horizontal scaling.

File: src/administration/horizontal_scaling.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+To solve this, Lemmy can be started with the `--disable-scheduled-tasks` flag on all but one instance. In general,
+there are two approaches:
+
+1. Run all your load balanced Lemmy servers with the `--disable-scheduled-tasks` flag, and run one additional Lemmy server
+   without this flag which is not in your load balancer and does not accept any HTTP traffic.
+2. Run one load balanced Lemmy server without the flag, and all other load balanced servers with the flag.
+
+Option 1 might have a few tiny advantages (easier to isolate logs for the scheduled tasks, and expensive scheduled tasks
+won't compete with HTTP requests for system resources), but requires a whole extra server.
```
Comment (Nutomic) [inline src/administration/horizontal_scaling.md:94]: > but requires a whole extra server

Server process (not a physical server)

File: src/administration/horizontal_scaling.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+#### Rolling upgrades
+
+For most versions, rolling releases have been completely OK, but there have been some cases where a database migration
+slips in which is NOT backwards compatible (causing any servers running the older version of Lemmy to potentially start
+generating errors). To be safe, you should always first take a look at what database migrations are included in a new
+version, and evaluate whether the changes are for sure safe to apply with running servers.
```
Comment (Nutomic) [inline src/administration/horizontal_scaling.md:101]: Link to migrations folder here.

File: src/administration/horizontal_scaling.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+Pict-rs is a service which does image processing. It handles user-uploaded images as well as downloading thumbnails for
+external images.
+
+Pict-rs does not scale horizontally as of writing this document. This is due to a dependency on a Sled database, which
+is a disk based database file.
+
+The developer of pict-rs has mentioned plans to eventually add Postgres support (which should in theory enable
+horizontal scaling), but work on this has not yet started.
+
+Note that by caching pict-rs images in a CDN, you can really minimize load on the pict-rs server!
```
Comment (Nutomic) [inline src/administration/horizontal_scaling.md:122]: Doesnt even need a CDN, you can do caching in nginx very effectively.

https://www.nginx.com/blog/nginx-caching-guide/

File: src/administration/horizontal_scaling.md (hunk 2 above)
Comment (sunaurus) [reply src/administration/horizontal_scaling.md:35]: >@phiresky I would always recommend turning on ip hash load leveling. that should prevent most of this?

It's similar to sticky sessions, I'll add it to the list 

> @nutomic Seems like the problem should be prevented by this change. Or is it not working?

That change is working without problems, but it doesn't help here - different lemmy-ui versions will have different hashes. So out of the box, if you request `/static/<hash_from_version_1>` from a lemmy-ui running version 2, you will get a 404 response.

Comment (sunaurus) [reply src/administration/horizontal_scaling.md:35]: I added the note about ip hash based load balancing

File: src/administration/horizontal_scaling.md (hunk 6 above)
Comment (sunaurus) [reply src/administration/horizontal_scaling.md]: I clarified this point

File: src/administration/horizontal_scaling.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+Lemmy-ui can be horizontally scaled, but make sure you read the following section for information about rolling upgrades.
+
+#### Rolling upgrades
+
+In general, there is nothing preventing you from running multiple load balanced Lemmy-ui servers in parallel, but
+without some custom configuration, **rolling upgrades will break Lemmy for your users!** This is because by default,
+Lemmy-ui will serve static files through the express process.
+
+Consider the scenario where you have 2 Lemmy-ui servers, one is running on version 1, and the other has just come online
+running version 2. A user opening your Lemmy in their browser might have their front page html request routed to version 1.
```
Comment (dessalines) [inline src/administration/horizontal_scaling.md:33]: Don't use code-style forced-line wrapping. Markdown can handle long lines fine.

Comment (sunaurus) [reply src/administration/horizontal_scaling.md]: Fixed

**Review by Nutomic — APPROVED** [review]

**Review by dessalines — APPROVED** [review]

**Pull request merged**